2.15.2.Final breaks command mode with main class extends from QuarkusApplication in kotlin

### Describe the bug

Same error as in #30054 .

```kotlin
import io.quarkus.runtime.QuarkusApplication
import io.quarkus.runtime.annotations.QuarkusMain

@QuarkusMain
class MainCLI : QuarkusApplication {

    @Throws(Exception::class)
    override fun run(vararg args: String): Int {
        return 0
    }
}
```

```bash
...
[ERROR] Failed to execute goal io.quarkus.platform:quarkus-maven-plugin:2.15.2.Final:build (default) on project code-with-quarkus: Failed to build quarkus application: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
[ERROR] 	[error]: Build step io.quarkus.deployment.steps.MainClassBuildStep#mainClassBuildStep threw an exception: java.lang.IllegalArgumentException: The supplied 'main-class' value of 'org.acme.MainCLIKt' does not correspond to either a fully qualified class name or a matching 'name' field of one of the '@QuarkusMain' annotations
...
```

### How to Reproduce?

[code-with-quarkus.zip](https://github.com/quarkusio/quarkus/files/10350129/code-with-quarkus.zip)

`./mvnw clean install`

### Output of `java -version`

Java 17

### Quarkus version or git rev

2.15.2.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Maven

--- a/core/deployment/src/main/java/io/quarkus/deployment/steps/MainClassBuildStep.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/steps/MainClassBuildStep.java
@@ -419,25 +419,15 @@ public class MainClassBuildStep {
     }
 
     private static String sanitizeMainClassName(ClassInfo mainClass, IndexView index) {
-        String className = mainClass.name().toString();
+        DotName mainClassDotName = mainClass.name();
+        String className = mainClassDotName.toString();
         if (isKotlinClass(mainClass)) {
             MethodInfo mainMethod = mainClass.method("main",
                     ArrayType.create(Type.create(DotName.createSimple(String.class.getName()), Type.Kind.CLASS), 1));
             if (mainMethod == null) {
-                ClassInfo classToCheck = mainClass;
-                boolean hasQuarkusApplicationSuperClass = false;
-                while (classToCheck != null) {
-                    DotName superName = classToCheck.superName();
-                    if (superName.equals(QUARKUS_APPLICATION)) {
-                        hasQuarkusApplicationSuperClass = true;
-                        break;
-                    }
-                    if (superName.equals(OBJECT)) {
-                        break;
-                    }
-                    classToCheck = index.getClassByName(superName);
-                }
-                if (!hasQuarkusApplicationSuperClass) {
+                boolean hasQuarkusApplicationInterface = index.getAllKnownImplementors(QUARKUS_APPLICATION).stream().map(
+                        ClassInfo::name).anyMatch(d -> d.equals(mainClassDotName));
+                if (!hasQuarkusApplicationInterface) {
                     className += "Kt";
                 }
             }

--- a/integration-tests/gradle/src/main/resources/kotlin-grpc-project/src/main/kotlin/org/acme/Main.kt
+++ b/integration-tests/gradle/src/main/resources/kotlin-grpc-project/src/main/kotlin/org/acme/Main.kt
@@ -1,0 +1,15 @@
+package org.acme
+
+import io.quarkus.runtime.Quarkus
+import io.quarkus.runtime.QuarkusApplication
+import io.quarkus.runtime.annotations.QuarkusMain
+
+@QuarkusMain
+class Main : QuarkusApplication {
+
+    @Throws(Exception::class)
+    override fun run(vararg args: String?): Int {
+        Quarkus.waitForExit()
+        return 0
+    }
+}

